database/api_test.go

```diff
@@ -112,15 +115,225 @@ func (matcher *queryMatcher) Matches(x interface{}) bool {
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.NoError(err)
+		require.Equal(0, total)
+	})
+
+	t.Run("total equal 1", func(t *testing.T) {
+		require := _require.New(t)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(1), nil)
+		api.EXPECT().Select(&target, query, record.NewListOptions(limit, offset)).Return(nil)
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.NoError(err)
+		require.Equal(1, total)
+	})
+
+	t.Run("error case: Count error", func(t *testing.T) {
+		require := _require.New(t)
+		expected := generator.String(20)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(0), errors.New(expected))
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.Equal(0, total)
+		require.EqualError(err, expected)
+	})
+
+	t.Run("error case: Select error", func(t *testing.T) {
+		require := _require.New(t)
+		expected := generator.String(20)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(1), nil)
+		api.EXPECT().Select(&target, query, record.NewListOptions(limit, offset)).Return(errors.New(expected))
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.EqualError(err, expected)
+		require.Equal(0, total)
+	})
+}
+
+func Test_Begin(t *testing.T) {
+	require := _require.New(t)
+	ctrl := gomock.NewController(t)
+	transaction := transaction.NewMockTransaction(ctrl)
+	var err error
+	api := &api{
+		tx:            transaction,
+		configuration: &InstanceConfig{MaxLimit: 100},
+	}
+	err = api.Begin()
+	require.NoError(err)
+}
+
+func Test_Rollback(t *testing.T) {
+	require := _require.New(t)
+	ctrl := gomock.NewController(t)
+	transaction := transaction.NewMockTransaction(ctrl)
+	var err error
+	api := &api{
+		tx:            transaction,
+		configuration: &InstanceConfig{MaxLimit: 100},
+	}
+
+	transaction.EXPECT().Rollback().Return(nil)
+
+	err = api.Rollback()
+
+	require.NoError(err)
+}
+
+func Test_Rollback_transactionIsNil(t *testing.T) {
+	require := _require.New(t)
+
+	api := &api{
+		tx:            nil,
+		configuration: &InstanceConfig{MaxLimit: 100},
+	}
+
+	err := api.Rollback()
+	require.EqualError(err, ErrMissingTransaction.Error())
+}
+
+func Test_Commit(t *testing.T) {
+	require := _require.New(t)
+	ctrl := gomock.NewController(t)
+	transaction := transaction.NewMockTransaction(ctrl)
+
+	var err error
+
+	api := &api{
+		tx:            transaction,
+		configuration: &InstanceConfig{MaxLimit: 100},
+	}
+
+	transaction.EXPECT().Commit().Return(nil)
+
+	err = api.Commit()
+
+	require.NoError(err)
+}
+
+func Test_Commit_transactionIsNil(t *testing.T) {
+	require := _require.New(t)
+
+	api := &api{
+		tx:            nil,
+		configuration: &InstanceConfig{MaxLimit: 100},
+	}
+
+	err := api.Commit()
+	require.EqualError(err, ErrMissingTransaction.Error())
+}
+
+func Test_ApiMethods(t *testing.T) {
```

Comment:
we will need another set of these without a transaction already in progress. This is only covering the case when a transaction is already in progress. You will probably need to create a mock to create the transactable you want to test against. If you have trouble with that ping @nmarie22 or I.